cirq-core/cirq/ops/gate_features.py

```diff
@@ -31,57 +32,19 @@ def qubit_index_to_equivalence_group_key(self, index: int) -> int:
         return 0
 
 
-class SupportsOnEachGate(raw_types.Gate, metaclass=abc.ABCMeta):
-    """A gate that can be applied to exactly one qubit."""
-
-    def on_each(self, *targets: Union[raw_types.Qid, Iterable[Any]]) -> List[raw_types.Operation]:
-        """Returns a list of operations applying the gate to all targets.
-        Args:
-            *targets: The qubits to apply this gate to. For single-qubit gates
-            this can be provided as varargs or a combination of nested
-            iterables. For multi-qubit gates this must be provided as an
-            `Iterable[Sequence[Qid]]`, where each sequence has `num_qubits`
-            qubits.
-        Returns:
-            Operations applying this gate to the target qubits.
-        Raises:
-            ValueError if targets are not instances of Qid or Iterable[Qid].
-            ValueError if the gate qubit number is incompatible.
-        """
-        operations: List[raw_types.Operation] = []
-        if self._num_qubits_() > 1:
-            iterator: Iterable = targets
-            if len(targets) == 1:
-                if not isinstance(targets[0], Iterable):
-                    raise TypeError(f'{targets[0]} object is not iterable.')
-                t0 = list(targets[0])
-                iterator = [t0] if t0 and isinstance(t0[0], raw_types.Qid) else t0
-            for target in iterator:
-                if not isinstance(target, Sequence):
-                    raise ValueError(
-                        f'Inputs to multi-qubit gates must be Sequence[Qid].'
-                        f' Type: {type(target)}'
-                    )
-                if not all(isinstance(x, raw_types.Qid) for x in target):
-                    raise ValueError(f'All values in sequence should be Qids, but got {target}')
-                if len(target) != self._num_qubits_():
-                    raise ValueError(f'Expected {self._num_qubits_()} qubits, got {target}')
-                operations.append(self.on(*target))
-            return operations
-
-        for target in targets:
-            if isinstance(target, raw_types.Qid):
-                operations.append(self.on(target))
-            elif isinstance(target, Iterable) and not isinstance(target, str):
-                operations.extend(self.on_each(*target))
-            else:
-                raise ValueError(
-                    f'Gate was called with type different than Qid. Type: {type(target)}'
-                )
-        return operations
-
-
-class SingleQubitGate(SupportsOnEachGate, metaclass=abc.ABCMeta):
+class _SupportsOnEachGateMeta(value.ABCMetaImplementAnyOneOf):
```

Comment:
can you remind me what this metaclass is doing? why couldn't we just have an empty `SupportsOnEachGate` class without this metaclass?

Reply:
We want to deprecate SupportsOnEachGate, but any gate that inherits from it will cause deprecation errors during unit tests. So we want to remove the inheritances. However for backwards compatibility we want isinstance(cirq.X, SupportsOnEachGate) to return True. The metaclass does this shimming for us.

Reply:
thanks for the explanation